stored method interference graph

vtjnash opened the pull request:
Locally, I measure about a 10-20% load time improvement for the packages in #57436, or about half of the measured regressions (and they hopefully should also run inference slightly faster too since all method lookup are now much faster). We'd hoped to stop computing this entirely before #265 was fixed, but since then it hasn't made a ton of sense that we have to compute most of this graph for identifying invalidations, and then we had to recompute it every time the user called `methods`. At this time it now seems to make more sense to just store the graph explicitly (it is fairly small size impact) and reap the performance benefits later. Most good code should already have been hitting the fast paths here and not see any impact from this, but most packages also contain at least some transitive dependency on something like ArrayLayouts (or BlockArrays or LazyArrays) whose implementation forces it onto the slow fallback paths for much of those packages code. In the future, this should also allow us to focus our performance efforts on the load performance of activating Method and better caching for those, since the load performance of `verify_call` is now relatively negligible.

(please preserve the first commit message not this description when merging)

vtjnash committed:
precompute all possible ml-matches lookup results

Store full method interference relationship graph in interferences field
of Method to avoid expensive morespecific calls during dispatch. This
provides significant performance improvements:
  - Replace method comparisons with precomputed interference lookup.
  - Optimize ml_matches minmax computation using interference lookups.
  - Optimize sort_mlmatches for large return sets by iterating over
    interferences instead of all matching methods.
  - Add method_morespecific_via_interferences in both C and Julia.

This representation may exclude some edges that are implied by
transitivity since sort_mlmatches will ensure the correct result by
following strong edges. Ambiguous edges are guaranteed to be checkable
without recursion.

Also fix a variety of bugs along the way:
 - Builtins signature would cause them to try to discard all other
   methods during `sort_mlmatches`.
 - Some ambiguities were over-estimated, which now are improved upon.
 - Setting lim==-1 now gives the same limited list of methods as lim>0,
   since that is actually faster now than attempting to give the
   unsorted list. This provides a better fix to #53814 than #57837 and
   fixes #58766.
 - Reverts recent METHOD_SIG_LATEST_HAS_NOTMORESPECIFIC attempt (though
   not the whole commit), since I found a significant problem with any
   usage of that bit during testing: it only tracks methods that
   intersect with a target, but new methods do not necessarily intersect
   with any existing target.

This provides a decent performance improvement to `methods` calls, which
implies a decent speed up to package loading also (e.g. ModelingToolkit
loads in about 4 seconds instead of 5 seconds).

vtjnash committed:
remove disabled code

Could restore this later, so keeping it around in the PR, but please squash merge this away.

KristofferC commented:
>  Most good code should already have been hitting the fast paths here and not see any impact from this, but most packages also contain at least some transitive dependency on something like ArrayLayouts (or BlockArrays or LazyArrays) whose implementation forces it onto the slow fallback paths

Out of curiosity, what bad patterns exist in those packages?

vtjnash commented:
Lots of LinearAlgebra-related overrides with excessively complicated Unions which usually map very poorly to the actual intent of the functions

vtjnash commented:
And lots of functions that fail to infer, resulting in extra work to verify no overloads have been added since

JeffBezanson commented on base/Base.jl:
```diff
@@ -267,7 +267,9 @@ include("uuid.jl")
 include("pkgid.jl")
 include("toml_parser.jl")
 include("linking.jl")
+module StaticData
```
Is there a reason to do this?

vtjnash commented on src/gf.c:
```diff
@@ -2621,6 +2623,133 @@ jl_typemap_entry_t *jl_method_table_add(jl_methtable_t *mt, jl_method_t *method,
+    arraylist_new(&seen, 0);
+    int result = method_in_interferences_recursive(start_method, target_method, &seen);
+    arraylist_free(&seen);
+    assert(result == jl_method_morespecific(target_method, start_method));
```
Looks like type intersect makes mistakes sometimes, which is something I assumed it would not do when adding an assert for typeintersect not making obvious mistakes.
```
julia> typeintersect(Tuple{Type{<:Base.BitInteger}}, Tuple{Type{Union{Missing, T}}} where T)
Tuple{Type{Union{Missing, T}}} where T # wrong

julia> typeintersect(Tuple{Type{Union{Missing, T}}} where T, Tuple{Type{<:Base.BitInteger}})
Union{} # correct
```

KristofferC commented:
https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/995ff9d_vs_228edd6/report.html suggests this needs some fixup.

vtjnash committed:
remove asserts

type-intersect makes mistakes, and does not need to crash the process when it does

KristofferC commented:
This runs vs 1.12 now which might give kind of noisy results.

KristofferC commented:
I think the report looks good (taking into account one running on master and the other on 1.12).

JeffBezanson commented on base/staticdata.jl:
```diff
@@ -282,51 +280,180 @@ function verify_method(codeinst::CodeInstance, stack::Vector{CodeInstance}, visi
+# Returns true if method2 is in method1's interferences (meaning !morespecific(method2, method1))
+function method_in_interferences_recursive(method2::Method, method1::Method, visited::IdSet{Method})
+    if method_in_interferences(method1, method2)
+        return false
```
This looks like it contradicts the comment; what am I missing?

vchuravy commented on base/staticdata.jl:
```diff
@@ -282,51 +280,180 @@ function verify_method(codeinst::CodeInstance, stack::Vector{CodeInstance}, visi
+
+# Check if method2 is in method1's interferences set
+# Returns true if method2 is found (meaning !morespecific(method1, method2))
+function method_in_interferences(method1::Method, method2::Method)
```
As a foreign speaker, the method name made my head hurt.

vtjnash committed:
review

IanButterworth commented:
Merge? (Noting the request to preserve the first commit message)

KristofferC commented:
> With JLLWrappers v1.7.1''

<img width="595" height="420" alt="image" src="https://github.com/user-attachments/assets/09998ea0-5f5c-4374-8966-2b011ce5ba8d" />

just joking
